pelican/log.py

```diff
@@ -1,142 +1,446 @@
-    _group_count = defaultdict(int)
+    def ignore(self, level, message):
+        self._ignored.add((level, message))
 
     def filter(self, record):
-        # don't limit log messages for anything above "warning"
-        if record.levelno > self.LOGS_DEDUP_MIN_LEVEL:
+        if record.levelno < self.level:
+            # Do not dedeuplicate messages of lower severity
             return True
 
-        # extract group
-        group = record.__dict__.get('limit_msg', None)
-        group_args = record.__dict__.get('limit_args', ())
+        # ignore duplicates of resolved/formatted messages:
+        message_key = (record.levelno, record.getMessage())  # resolved msg
+        ignored = self._ignored
 
-        # ignore record if it was already raised
-        message_key = (record.levelno, record.getMessage())
-        if message_key in self._raised_messages:
-            return False
+        if message_key in ignored:
+            return False  # skip messages already encountered
         else:
-            self._raised_messages.add(message_key)
-
-        # ignore LOG_FILTER records by templates or messages
-        # when "debug" isn't enabled
-        logger_level = logging.getLogger().getEffectiveLevel()
-        if logger_level > logging.DEBUG:
-            template_key = (record.levelno, record.msg)
-            message_key = (record.levelno, record.getMessage())
-            if (template_key in self._ignore or message_key in self._ignore):
-                return False
-
-        # check if we went over threshold
-        if group:
-            key = (record.levelno, group)
-            self._group_count[key] += 1
-            if self._group_count[key] == self._threshold:
-                record.msg = group
-                record.args = group_args
-            elif self._group_count[key] > self._threshold:
-                return False
+            ignored.add(message_key)  # continue, but mark to ignore next time
+
         return True
 
 
-class LimitLogger(logging.Logger):
+class BlacklistFilter(logging.Filter):
+    """Use to prevent messages of specific level & content from emitting."""
+
+    def __init__(self, item_pairs=()):
+        super().__init__()
+        self._ignored = set()
+        for lvl, msg in item_pairs:
+            self.ignore(lvl, msg)
+
+    @classmethod
+    def get_filters(cls, logger):
+        """Return a list of BlacklistFilters currently on logger."""
+        found = []
+        for x in logger.filters:
+            if isinstance(x, cls):
+                found.append(x)
+        return found
+
+    def ignore(self, level, message):
+        self._ignored.add((level, message))
+
+    def filter(self, record):
+        if not self._ignored:
+            return True
+        if (
+            ((record.levelno, record.msg) in self._ignored)
+            or
+            ((record.levelno, record.getMessage()) in self._ignored)
+        ):
+            return False  # do not emit msg
```

Comment:
I agree, this could be improved with comments.

However I prefer the boolean logic over two if-statements:  the result is to follow the same logic branch in both cases so I'd rather not create a new branch.  I'll take another stab at this to see if I can make it more clear without adding more branching.